### ISSUE TITLE
🧠 Model | Replace Softmax with PowerSoftmax in Phi-3-Mini-4K-Instruct

### ISSUE THREAD

**Issue by collinbentley1** [medlock-ai/medlock#26]
### Goal
Produce `phi3-powersoftmax.pt`—a drop-in checkpoint where every attention head uses PowerSoftmax (p = 4, ε = 1e-3) and all forward/backward paths are parity-tested.

### Codebase Layout
```
medlock_fhe/
  layers/
    power_softmax.py
  models/
    phi3_power/
      modeling_phi3_power.py
  tests/
    test_power_softmax.py
    test_phi3_power_perplexity.py
```

### Detailed Tasks
1. **Implement PowerSoftmax layer**
   * Accept `x: Tensor[B, T, H]`; compute `x_clamped = torch.relu(x)` (values ≥ 0).
   * `numer = x_clamped.pow(p)`
   * `denom = numer.sum(dim=-1, keepdim=True).add_(epsilon)`
   * Return `numer / denom`
   * Provide custom `autograd.Function` if fused kernel unreachable; else fallback to eager.
   * Unit-gradient test: 
     ```python
     torch.autograd.gradcheck(PowerSoftmax(...), ...)
     ```
2. **Fork HF modelling file**
   * Sub-class `PhiAttention` → `PhiPowerAttention`; swap `nn.Softmax(dim=-1)` with `PowerSoftmax`.
   * Preserve weight names so HF loaders work.
3. **Checkpoint conversion script**
   ```
   python scripts/convert_phi3_to_power.py \
     --src hf-internal-testing/phi-3-mini-4k-instruct \
     --dst ./weights/phi3_power
   ```
   * Iterate `state_dict`; nothing changes except no longer referencing Softmax in config.
4. **Quick Perplexity Smoke Test**
   * Use `datasets.load_dataset("wikitext", "wikitext-2-raw-v1")`.
   * Batched evaluation (max_seq_len = 512).
   * Compute ppl for baseline vs PowerSoftmax; assert Δ ≤ +0.15 on log-perplexity.

5. **Performance Micro-bench**
   * Introduce `bench/softmax_vs_powersoftmax.py`:
     * Runs forward pass on A100 for tensor shape (B=64, H=32, T=128).
     * Record CUDA events; target ≤ 1.4× baseline Softmax latency.

6. **CI Integration**
   * Add `pytest -q tests/test_power_softmax.py` in main workflow.
   * Block PR if perplexity delta > 15 %.

### Acceptance Criteria
- `phi3_power/config.json` identical to upstream except `"attention_impl": "power_softmax"`.
- `scripts/convert...` finishes under 2 min CPU.
- `pytest` passes incl. gradient parity (mean abs diff < 1e-3).
- `bench` script prints both timings; PowerSoftmax within 40 % of Softmax.
